modules/order/src/AvailabilityOrderProcessor.php

```diff
@@ -37,10 +38,23 @@ public function process(OrderInterface $order) {
     foreach ($order->getItems() as $order_item) {
       $purchased_entity = $order_item->getPurchasedEntity();
       if ($purchased_entity) {
-        $available = $this->availabilityManager->check($purchased_entity, $order_item->getQuantity(), $context);
-        if (!$available) {
-          $order->removeItem($order_item);
-          $order_item->delete();
+        /** @var \Drupal\commerce\AvailabilityResponseInterface $availability */
+        $availability = $this->availabilityManager->check($purchased_entity, $order_item->getQuantity(), $context);
+        if (!$availability->isNeutral()) {
+          if ($availability->getMax() == 0) {
+            $order->removeItem($order_item);
+            $order_item->delete();
+            drupal_set_message(t('The item %item is no longer available and has been removed from your cart.', [
+              '%item' => $order_item->getTitle(),
+            ]), 'warning');
+          }
+          elseif ($availability->getMax() < $order_item->getQuantity()) {
+            drupal_set_message(new PluralTranslatableMarkup(abs($availability->getMax() - $order_item->getQuantity()), '1 %item is currently unavailable. The quantity in your cart has been updated.', '@count %item are currently unavailable. The quantity in your cart has been updated.', [
```

Comment:
This should use the "availability Reason" something like:
```
            drupal_set_message(new PluralTranslatableMarkup(abs($availability->getMax() - $order_item->getQuantity()), '1 %item %reason. The quantity in your cart has been updated.', '@count %item %reason. The quantity in your cart has been updated.', [
              '%item' => $order_item->getTitle(), '%reason' => $availability->getReason(),
            ]), 'warning');
```
Also the part: "The quantity in your cart has been updated" is not always true as the quantity may have stayed the same but that's less of an issue.